wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointPlugin.java

```diff
@@ -0,0 +1,370 @@
+        TimeUnit.MILLISECONDS.toNanos(this.idleMonitorExpirationMs)
+    );
+  }
+
+  /**
+   * Determines the AWS region from the given parameters. The region specified by the props will be used if it is
+   * provided. Otherwise, the region will be parsed from the specified host URL.
+   *
+   * @param hostSpec Host information for the connection being established.
+   * @param props    The connection properties for the connection being established.
+   * @return The AWS region of the connection being established.
+   */
+  protected Region getRegion(HostSpec hostSpec, Properties props) {
+    String regionString = REGION_PROPERTY.getString(props);
+    if (StringUtils.isNullOrEmpty(regionString)) {
+      regionString = rdsUtils.getRdsRegion(hostSpec.getHost());
+    }
+
+    if (StringUtils.isNullOrEmpty(regionString)) {
+      throw new
+          RuntimeException(
+          Messages.get(
+              "CustomEndpointMonitorImpl.missingRequiredConfigParameter",
+              new Object[] {REGION_PROPERTY.name}));
+    }
+
+    final Region region = Region.of(regionString);
+    if (!Region.regions().contains(region)) {
+      throw new RuntimeException(
+          Messages.get(
+              "AwsSdk.unsupportedRegion",
+              new Object[] {regionString}));
+    }
+
+    return region;
+  }
+
+  /**
+   * If custom endpoint info does not exist for the current custom endpoint, waits a short time for the info to be
+   * made available by the custom endpoint monitor. This is necessary so that other plugins can rely on accurate custom
+   * endpoint info. Since custom endpoint monitors and information are shared, we should not have to wait often.
+   *
+   * @return true if the custom endpoint info is available, or false if we timed out while waiting for the info to
+   *     become available.
+   */
+  protected boolean waitForCustomEndpointInfo() {
+    CustomEndpointInfo cachedInfo =
+        this.pluginService.getInfo(this.customEndpointHostSpec.getHost(), CustomEndpointInfo.class, true);
+    boolean isInfoInCache = cachedInfo != null;
+
+    if (!isInfoInCache) {
+      // Wait for the monitor to place the custom endpoint info in the cache. This ensures other plugins get accurate
+      // custom endpoint info.
+      this.waitForInfoCounter.inc();
+      LOGGER.fine(
+          Messages.get(
+              "CustomEndpointPlugin.waitingForCustomEndpointInfo",
+              new Object[]{ this.customEndpointHostSpec.getHost(), this.waitOnCachedInfoDurationMs }));
+      long waitForEndpointInfoTimeoutNano =
+          System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(this.waitOnCachedInfoDurationMs);
+
+      while (!isInfoInCache && System.nanoTime() < waitForEndpointInfoTimeoutNano) {
+        cachedInfo =
+            this.pluginService.getInfo(this.customEndpointHostSpec.getHost(), CustomEndpointInfo.class, true);
+        isInfoInCache = cachedInfo != null;
+      }
+
+      if (isInfoInCache) {
+        LOGGER.fine(
```

Comment:
Should we raise TimeoutException here?

Reply:
I will change it to throw a SQLException if we timed out and there is not info in the cache. I cannot make it a TimeoutException because this method is called from connect and execute and both places do not have TimeoutException in their method signature.